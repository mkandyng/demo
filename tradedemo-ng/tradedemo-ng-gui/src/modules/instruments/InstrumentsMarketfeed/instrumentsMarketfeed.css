#instrumentsMarketfeed .item {
  padding: 2px 6px;
  cursor: default;
}

#instrumentsMarketfeed .item-header {
  background-color: #eeeeee;
  color: #454545;
  font-weight: bold;
}

#instrumentsMarketfeed .ReactTable .rt-thead {
  height:30px;
}

#instrumentsMarketfeed .ReactTable {
  border-radius: 5px;
  background: #fffdfd;
  font-size: 10px;
  text-align: center;
  height: 270px;
}

#instrumentsMarketfeed .ReactTable .rt-th {
  font-weight: bold;
  background: #f7f7f7;
}

#instrumentsMarketfeed .ReactTable .rt-tr-group:hover {
  background: #f7f7f7;
}

#instrumentsMarketfeed .ReactTable .rt-td-group {
  background: #ece5e5;
}

#instrumentsMarketfeed .ReactTable .instrument-column {
  text-align: left;
}

#instrumentsMarketfeed .ReactTable span.priceUp {
  color: blue;
}

#instrumentsMarketfeed .ReactTable span.priceDown {
  color: red;
}
