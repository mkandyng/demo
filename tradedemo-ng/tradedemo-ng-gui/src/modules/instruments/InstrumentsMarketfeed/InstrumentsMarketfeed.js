import React from "react";
import ReactTable from "react-table";
import {MAX_MARKET_FEED_INSTRUMENTS, MOUSE_OVER_IMAGE, MOUSE_OUT_IMAGE} from "../instruments";
import PropTypes from 'prop-types';
import "./instrumentsMarketfeed.css";

export const SELECTED_MARKET_FEED_ROW_COLOUR = "#f5f2f2";

/**
 * [InstrumentsMarketfeed Component to handle Marketfeed view and interaction]
 * @param       {[Array]} marketfeedInstruments      [store instruments.marketfeedInstruments]
 * @param       {[Array]} instrument                 [store instruments.selected]
 * @param       {[Function]} selectMarketfeedInstrument [action to select an instrument in the marketfeed]
 * @param       {[Function]} deleteMarketfeedInstrument [action to delete an instrument in the marketfeed]
 * @param       {[Function]} updateMarketfeedInstrument [action to update an instrument in the marketfeed]
 * @constructor
 */

export default function InstrumentsMarketfeed({
  marketfeedInstruments,
  instrument,
  selectMarketfeedInstrument,
  deleteMarketfeedInstrument,
  updateMarketfeedInstrument
}) {
  const eventHandlers = {
    handleTableColumn: (state, rowInfo, column) => {
      return handleTableColumn(state,
                               rowInfo,
                               column,
                               marketfeedInstruments,
                               deleteMarketfeedInstrument,
                               selectMarketfeedInstrument);
    },
    handleTableRow: (state, rowInfo) => {
      return handleTableRow(state,
                            rowInfo,
                            instrument,
                            marketfeedInstruments,
                            selectMarketfeedInstrument);
    }
  };

  return (<InstrumentsMarketfeedView
                            instruments={marketfeedInstruments}
                            columns={columnHeaders}
                            eventHandlers={eventHandlers}/>);
}

function InstrumentsMarketfeedView({instruments, columns, eventHandlers}) {
  return (<ReactTable
                defaultPageSize={MAX_MARKET_FEED_INSTRUMENTS}
                sortable={false}
                showPaginationTop={false}
                showPaginationBottom={false}
                showPageSizeOptions={false}
                data={instruments}
                columns={columns}
                getTdProps={eventHandlers.handleTableColumn}
                getTrProps={eventHandlers.handleTableRow}/>);
}

export const columnHeaders = [
  {
    Header: "Remove",
    accessor: "delete",
    width: 70
  }, {
    Header: "Symbol",
    accessor: "symbol",
    width: 100
  }, {
    Header: "Instrument Name",
    accessor: "name",
    className: "instrument-column"
  }, {
    Header: "Ccy",
    accessor: "currency",
    width: 70
  }, {
    Header: "Bid",
    accessor: "bid",
    width: 70
  }, {
    Header: "Ask",
    accessor: "ask",
    width: 70
  }, {
    Header: "Last",
    accessor: "last",
    width: 70
  }, {
    Header: "Open",
    accessor: "open",
    width: 70
  }, {
    Header: "Chg",
    accessor: "chg",
    width: 70
  }
];

function removeColumnHeader(column) {
  return column.Header === "Remove";
}

function selectMarketfeed(symbol, marketfeedInstruments, selectMarketfeedInstrument) {
  const matchedInstrument = marketfeedInstruments.find(instrument => instrument.symbol === symbol);
  selectMarketfeedInstrument(matchedInstrument);
}

function handleTableRow(state, rowInfo, instrument, marketfeedInstruments, selectMarketfeedInstrument) {
  const index = rowInfo? rowInfo.index: -1;
  const matchedPredicate = e => e.symbol === instrument.symbol;
  const selectedIndex = marketfeedInstruments.findIndex(matchedPredicate);
  return {
    onClick: (e) => {
      if (e.target.src === undefined || !e.target.src.indexOf("delete")) {
        const symbol = marketfeedInstruments[index].symbol;
        if (symbol !== instrument.symbol) {
          selectMarketfeed(symbol, marketfeedInstruments, selectMarketfeedInstrument);
        }
      }
    },
    style: {
      background: selectedIndex === index? SELECTED_MARKET_FEED_ROW_COLOUR: null
    }
  };
};

function handleTableColumn(state,
                           rowInfo,
                           column,
                           marketfeedInstruments,
                           deleteMarketfeedInstrument,
                           selectMarketfeedInstrument) {
  const index = rowInfo? rowInfo.index: -1;
  return {
    onClick: (e) => {
      if (removeColumnHeader(column)) {
        const instrument = {
          symbol: rowInfo.original.symbol,
          name: rowInfo.original.name,
          currency: rowInfo.original.currency
        }
        if (marketfeedInstruments.length > 1) {
          const nextSymbol = index === 0? marketfeedInstruments[1].symbol: marketfeedInstruments[0].symbol;
          deleteMarketfeedInstrument(instrument);
          selectMarketfeed(nextSymbol, marketfeedInstruments, selectMarketfeedInstrument);
        } else {
          alert("Must have at least one instrument in the market feed");
        }
      }
    },
    onMouseOver: (e) => {
      if (removeColumnHeader(column)) {
        e.target.src = MOUSE_OVER_IMAGE;
      }
    },
    onMouseOut: (e) => {
      if (removeColumnHeader(column)) {
        e.target.src = MOUSE_OUT_IMAGE;
      }
    }
  };
};

InstrumentsMarketfeed.propTypes = {
  marketfeedInstruments: PropTypes.arrayOf(PropTypes.object).isRequired,
  instrument: PropTypes.object.isRequired,
  selectMarketfeedInstrument: PropTypes.func.isRequired,
  deleteMarketfeedInstrument: PropTypes.func.isRequired,
  updateMarketfeedInstrument: PropTypes.func.isRequired
};

InstrumentsMarketfeedView.propTypes = {
  instruments: PropTypes.arrayOf(PropTypes.object).isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  eventHandlers: PropTypes.objectOf(PropTypes.func).isRequired
};
