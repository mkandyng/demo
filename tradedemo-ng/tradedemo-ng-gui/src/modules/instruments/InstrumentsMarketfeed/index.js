import InstrumentsMarketfeed from "./InstrumentsMarketfeed";

export default InstrumentsMarketfeed;
