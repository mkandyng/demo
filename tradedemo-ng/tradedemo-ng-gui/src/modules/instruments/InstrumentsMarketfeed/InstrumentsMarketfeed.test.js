import React from "react";
import toJson from "enzyme-to-json";
import { shallow } from "enzyme";
import ReactTable from "react-table";
import InstrumentsMarketfeed, { InstrumentsMarketfeedView, columnHeaders } from "./InstrumentsMarketfeed";

/**
 * This is unit test of Marketfeed component
 * It leverage snapshot to test the presentation
 */

describe("InstrumentsMarketfeed", () => {
    const props = {
        marketfeedInstruments: [ { delete:<img src="img/delete.png" alt="delete" />,
                                       symbol: "AMZ" } ],
        instrument: {symbol: "AMZ"},
        selectMarketfeedInstrument: jest.fn(),
        deleteMarketfeedInstrument: jest.fn(),
        updateMarketfeedInstrument: jest.fn()
    };

    it("should render component comparing with previous snapshot shallow markup", () => {
        // Given

        // When
        const component = shallow(<InstrumentsMarketfeed {...props} />);

        // Then
        expect(toJson(component)).toMatchSnapshot();
    });

    it("should registered event handler correctly", () => {
        // Given
        const localProps = {
            instruments: props.marketfeedInstruments,
            columns: columnHeaders,
            eventHandlers: {
                handleTableColumn: jest.fn(),
                handleTableRow: jest.fn()
            }
        };

        // When
        const component = shallow(<InstrumentsMarketfeedView {...localProps} />);

        // Then
        const reactTable = component.find(ReactTable);
        expect(reactTable.find({ getTdProps: localProps.eventHandlers.handleTableColumn })
                         .exists()).toBeTruthy();
        expect(reactTable.find({ getTrProps: localProps.eventHandlers.handleTableRow })
                         .exists()).toBeTruthy();
    });
});
