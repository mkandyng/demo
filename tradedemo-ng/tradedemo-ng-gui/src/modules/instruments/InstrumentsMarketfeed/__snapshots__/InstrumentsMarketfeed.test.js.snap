// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstrumentsMarketfeed should render component comparing with previous snapshot shallow markup 1`] = `
<ReactTable
  AggregatedComponent={[Function]}
  ExpanderComponent={[Function]}
  FilterComponent={[Function]}
  LoadingComponent={[Function]}
  NoDataComponent={[Function]}
  PadRowComponent={[Function]}
  PaginationComponent={[Function]}
  PivotValueComponent={[Function]}
  ResizerComponent={[Function]}
  TableComponent={[Function]}
  TbodyComponent={[Function]}
  TdComponent={[Function]}
  TfootComponent={[Function]}
  ThComponent={[Function]}
  TheadComponent={[Function]}
  TrComponent={[Function]}
  TrGroupComponent={[Function]}
  aggregatedKey="_aggregated"
  className=""
  collapseOnDataChange={true}
  collapseOnPageChange={true}
  collapseOnSortingChange={true}
  column={
    Object {
      "Aggregated": undefined,
      "Cell": undefined,
      "Expander": undefined,
      "Filter": undefined,
      "Footer": undefined,
      "Header": undefined,
      "Pivot": undefined,
      "PivotValue": undefined,
      "Placeholder": undefined,
      "aggregate": undefined,
      "className": "",
      "filterAll": false,
      "filterMethod": undefined,
      "filterable": undefined,
      "footerClassName": "",
      "footerStyle": Object {},
      "getFooterProps": [Function],
      "getHeaderProps": [Function],
      "getProps": [Function],
      "headerClassName": "",
      "headerStyle": Object {},
      "minResizeWidth": 11,
      "minWidth": 100,
      "resizable": undefined,
      "show": true,
      "sortMethod": undefined,
      "sortable": undefined,
      "style": Object {},
    }
  }
  columns={
    Array [
      Object {
        "Header": "Remove",
        "accessor": "delete",
        "width": 70,
      },
      Object {
        "Header": "Symbol",
        "accessor": "symbol",
        "width": 100,
      },
      Object {
        "Header": "Instrument Name",
        "accessor": "name",
        "className": "instrument-column",
      },
      Object {
        "Header": "Ccy",
        "accessor": "currency",
        "width": 70,
      },
      Object {
        "Header": "Bid",
        "accessor": "bid",
        "width": 70,
      },
      Object {
        "Header": "Ask",
        "accessor": "ask",
        "width": 70,
      },
      Object {
        "Header": "Last",
        "accessor": "last",
        "width": 70,
      },
      Object {
        "Header": "Open",
        "accessor": "open",
        "width": 70,
      },
      Object {
        "Header": "Chg",
        "accessor": "chg",
        "width": 70,
      },
    ]
  }
  data={
    Array [
      Object {
        "delete": <img
          alt="delete"
          id="deleteImg-AMZ"
          src="img/delete.png"
        />,
        "symbol": "AMZ",
      },
      Object {
        "delete": <img
          alt="delete"
          id="deleteImg-ABC"
          src="img/delete.png"
        />,
        "symbol": "ABC",
      },
    ]
  }
  defaultExpanded={Object {}}
  defaultFilterMethod={[Function]}
  defaultFiltered={Array []}
  defaultPage={0}
  defaultPageSize={5}
  defaultResized={Array []}
  defaultSortDesc={false}
  defaultSortMethod={[Function]}
  defaultSorted={Array []}
  expanderDefaults={
    Object {
      "filterable": false,
      "resizable": false,
      "sortable": false,
      "width": 35,
    }
  }
  filterable={false}
  freezeWhenExpanded={false}
  getLoadingProps={[Function]}
  getNoDataProps={[Function]}
  getPaginationProps={[Function]}
  getProps={[Function]}
  getResizerProps={[Function]}
  getTableProps={[Function]}
  getTbodyProps={[Function]}
  getTdProps={[Function]}
  getTfootProps={[Function]}
  getTfootTdProps={[Function]}
  getTfootTrProps={[Function]}
  getTheadFilterProps={[Function]}
  getTheadFilterThProps={[Function]}
  getTheadFilterTrProps={[Function]}
  getTheadGroupProps={[Function]}
  getTheadGroupThProps={[Function]}
  getTheadGroupTrProps={[Function]}
  getTheadProps={[Function]}
  getTheadThProps={[Function]}
  getTheadTrProps={[Function]}
  getTrGroupProps={[Function]}
  getTrProps={[Function]}
  groupedByPivotKey="_groupedByPivot"
  indexKey="_index"
  loading={false}
  loadingText="Loading..."
  multiSort={true}
  nestingLevelKey="_nestingLevel"
  nextText="Next"
  noDataText="No rows found"
  ofText="of"
  onFetchData={[Function]}
  originalKey="_original"
  pageJumpText="jump to page"
  pageSizeOptions={
    Array [
      5,
      10,
      20,
      25,
      50,
      100,
    ]
  }
  pageText="Page"
  pivotDefaults={Object {}}
  pivotIDKey="_pivotID"
  pivotValKey="_pivotVal"
  previousText="Previous"
  resizable={true}
  resolveData={[Function]}
  rowsSelectorText="rows per page"
  rowsText="rows"
  showPageJump={true}
  showPageSizeOptions={false}
  showPagination={true}
  showPaginationBottom={false}
  showPaginationTop={false}
  sortable={false}
  style={Object {}}
  subRowsKey="_subRows"
/>
`;
